src/object/obj_enum.c

```diff
@@ -262,6 +265,22 @@ fill_key(daos_handle_t ih, vos_iter_entry_t *key_ent, struct dss_enum_arg *arg,
 
 	iov = &arg->sgl->sg_iovs[arg->sgl_idx];
 
+	if (type == OBJ_ITER_DKEY && key_ent->ie_obj_punch && arg->need_punch &&
+	    !arg->obj_punched) {
+		int pi_size = sizeof(key_ent->ie_obj_punch);
+
+		arg->kds[arg->kds_len].kd_key_len = pi_size;
+		arg->kds[arg->kds_len].kd_val_type = OBJ_ITER_OBJ_PUNCH_EPOCH;
+		arg->kds_len++;
+
+		D_ASSERT(iov->iov_len + pi_size < iov->iov_buf_len);
+		memcpy(iov->iov_buf + iov->iov_len, &key_ent->ie_obj_punch,
+		       pi_size);
+
+		iov->iov_len += pi_size;
+		arg->obj_punched = true;
+	}
+
```

Comment:
I thought we will get object punch epoch by scan. see rebuild_obj_scan_cb() L601.  So you do not need change enumerate packing.  
Anything concern?

Reply:
I discussed this option with Jeff.  He says the scan only picks up the latest punch.  For the incarnation log problem we're trying to solve, the subsequent punch closest to the update epoch must me replayed.  There can be multiple object punches at differing epochs for an object, and recording them in the enumeration packing is required.

Reply:
The latest isn't sufficient.  You could have many snapshots you are rebuilding.   As such, you could have something like this

update object at time 1
take snapshot at 2
punch at 3
update a at 4
snapshot at 5
punch at 6

If you rebuild the object at 7, you will only see the punch at 6 in the scan phase.   You only see the intermediate punch 3 while you are doing migration for the snapshot at 2.   So you can't do this at the scan phase.

Reply:
ah, right. Thanks for explanation.